internal/ceb/child_exec.go

```diff
@@ -56,6 +163,66 @@ func (ceb *CEB) execChildCmd(ctx context.Context) <-chan error {
 	return ch
 }
 
+// termChildCmd terminates the child command.
+//
+// If force is set to true, this will send a SIGKILL.
+//
+// If force is false, this will send a SIGTERM and wait up to 30 seconds
+// for the child process to gracefully exit. If the process does not gracefully
+// exit in 30 seconds, we will send a SIGKILL.
+func (ceb *CEB) termChildCmd(
+	log hclog.Logger,
+	cmd *exec.Cmd,
+	childErrCh <-chan error, // error channel from startChildCmd
+	force bool,
+) error {
+	log = log.With("pid", cmd.Process.Pid)
+
+	// If we're not forcing, try a SIGTERM first.
+	if !force {
+		log.Info("sending SIGTERM")
+		if err := cmd.Process.Signal(syscall.SIGTERM); err != nil {
+			log.Warn("error sending SIGTERM, will proceed to SIGKILL", "err", err)
+		} else {
+			log.Debug("SIGTERM sent, waiting for child process to end or timeout")
+			select {
+			case err := <-childErrCh:
+				// If we got an exit error then everything worked propertly so
+				// we just set error to nil.
+				if _, ok := err.(*exec.ExitError); ok {
+					err = nil
+				}
+
+				// Child successfully exited.
+				log.Info("child process exited", "wait_err", err)
+				return err
+
+			case <-time.After(30 * time.Second):
+				// Timeout, fall through to SIGKILL
+				log.Warn("graceful termination failed, will send SIGKILL")
+			}
+		}
+	}
+
+	// SIGKILL
+	log.Info("sending SIGKILL")
+	if err := syscall.Kill(-cmd.Process.Pid, syscall.SIGKILL); err != nil {
```

Comment:
Worth a comment here to indicated why we're sending the negative pid value. I believe it's because you want to send it to the whole process group.